network-mux/test/Test/Mux.hs

```diff
@@ -1270,6 +1317,152 @@ prop_mux_start_mX apps runTime = do
+                           mux
+                          (daNum $ fst app)
+                          InitiatorDirectionOnly
+                          StartEagerly
+                          (dummyRestartingAppToChannel app)
+                       | app <- apps
+                       ]
+    r <- runRestartingApps mux $ M.fromList $ zip (map (daNum . fst) apps) getRes
+    stopMux mux
+    void $ waitCatch mux_aid
+    return $ property r
+
+  where
+    runRestartingApps :: Mux InitiatorMode m
+                      -> M.Map MiniProtocolNum (STM m (Either SomeException (DummyApp, Int)))
+                      -> m Bool
+    runRestartingApps _ ops | M.null ops = return True
+    runRestartingApps mux ops = do
+        appResult <- atomically $ foldr (<|>) retry $ M.elems ops
+        case appResult of
+             Left _ -> return False
+             Right (app, 0) -> do
+                 runRestartingApps mux $ M.delete (daNum app) ops
+             Right (app, restarts) -> do
+                 op <- runMiniProtocol mux (daNum app) InitiatorDirectionOnly StartEagerly
+                         (dummyRestartingAppToChannel (app, restarts - 1))
+                 runRestartingApps mux $ M.insert (daNum app) op ops
+
+prop_mux_restart_m (DummyRestartingResponderApps rapps) = do
+    mux_w <- atomically $ newTBQueue 10
+    mux_r <- atomically $ newTBQueue 10
+    let bearer = queuesAsMuxBearer nullTracer mux_w mux_r 1234
+        peerBearer = queuesAsMuxBearer nullTracer mux_r mux_w 1234
+        apps = map fst rapps
+        MiniProtocolBundle minis = MiniProtocolBundle $ map (appToInfo ResponderDirectionOnly) apps
+
+    mux <- newMux $ MiniProtocolBundle minis
+    mux_aid <- async $ runMux nullTracer mux bearer
+    getRes <- sequence [ runMiniProtocol
+                           mux
+                          (daNum $ fst app)
+                          ResponderDirectionOnly
+                          StartEagerly
+                          (dummyRestartingAppToChannel app)
+                       | app <- rapps
+                       ]
+    triggers <- mapM (async . (triggerApp peerBearer)) apps
+    r <- runRestartingApps mux $ M.fromList $ zip (map daNum apps) getRes
+    stopMux mux
+    void $ waitCatch mux_aid
+    mapM_ cancel triggers
+    return $ property r
+  where
+    runRestartingApps :: Mux ResponderMode m
+                      -> M.Map MiniProtocolNum (STM m (Either SomeException (DummyApp, Int)))
+                      -> m Bool
+    runRestartingApps _ ops | M.null ops = return True
+    runRestartingApps mux ops = do
+        appResult <- atomically $ foldr (<|>) retry $ M.elems ops
+        case appResult of
+             Left _ -> return False
+             Right (app, 0) -> do
+                 runRestartingApps mux $ M.delete (daNum app) ops
+             Right (app, restarts) -> do
+                 op <- runMiniProtocol mux (daNum app) ResponderDirectionOnly StartOnDemand
```

Comment:
If we add a delay before call to `runMiniProtocol` we can check how restarting works if there are outstanding bytes in the ingress queue.

Reply:
There are always bytes in the ingress queue in this test case.